File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            int forSum = -(nums[i]);
            int j = i + 1, k = nums.length - 1;
            while (j < k) {
                if (nums[j] + nums[k] == forSum) {
                    List<Integer> pair = Arrays.asList(nums[i], nums[j], nums[k]);
                    ans.add(pair);
                    j++;
                    k--;
                    while (j < k && nums[j] == nums[j - 1]) j++;
                    while (j < k && nums[k] == nums[k + 1]) k--;
                }
                else if (nums[j] + nums[k] < forSum)
                    j++;
                else
                    k--;
            }
        }
        return ans;
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        ans = set()

        for i in range(0, len(nums)):
            hs = set()
            if i > 0 and nums[i-1] == nums[i]:
                continue
            for j in range(i+1, len(nums)):
                if -(nums[i] + nums[j]) in hs:
                    trip = sorted([nums[i], nums[j], -(nums[i]+nums[j])])
                    ans.add(tuple(trip))
                else:
                    hs.add(nums[j])
        
        return [list(t) for t in ans]